ValidatingAdmissionPolicy causes kube-controller-manager to panic

### What happened?

I have a validatingadmissionpolicy which appears to cause kube-controller-manager to panic on a nil pointer dereference. This is the backtrace from the log:

```
E0123 12:33:29.270214       1 panic.go:262] "Observed a panic" panic="runtime error: invalid memory address or nil pointer dereference" panicGoValue="\"invalid memory address or nil pointer dereference\"" stacktrace=<
        goroutine 3482 [running]:
        k8s.io/apimachinery/pkg/util/runtime.logPanic({0x38d6f30, 0x56429e0}, {0x2e10c60, 0x55c0f10})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:107 +0xbc
        k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x38d6f30, 0x56429e0}, {0x2e10c60, 0x55c0f10}, {0x56429e0, 0x0, 0x10000000043ab05?})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:82 +0x5e
        k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00378d180?})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:59 +0x108
        panic({0x2e10c60?, 0x55c0f10?})
                runtime/panic.go:791 +0x132
        k8s.io/apiserver/pkg/cel.(*DeclType).IsObject(...)
                k8s.io/apiserver/pkg/cel/types.go:223
        k8s.io/apiserver/pkg/cel.(*DeclType).MaybeAssignTypeName(0x0, {0xc003f12b90?, 0x1?})
                k8s.io/apiserver/pkg/cel/types.go:117 +0x2d
        k8s.io/apiserver/pkg/admission/plugin/policy/validating.(*TypeChecker).declType(0xc001c46ba0?, {{0xc0029b4930, 0x11}, {0xc000daabb7, 0x7}, {0xc002887a38, 0x6}})
                k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go:253 +0x146
        k8s.io/apiserver/pkg/admission/plugin/policy/validating.(*TypeChecker).CreateContext(0xc001c46ba0, 0xc002bf8000)
                k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go:149 +0x18f
        k8s.io/apiserver/pkg/admission/plugin/policy/validating.(*TypeChecker).Check(0xc001c46ba0, 0xc002bf8000)
                k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go:111 +0x32
        k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus.(*Controller).reconcile(0xc001c58380, {0x38d70d8, 0xc001c0ee40}, 0xc002bf8000)
                k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus/controller.go:147 +0x78
        k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus.(*Controller).processNextWorkItem.func1(0xc001c58380, {0xc0029a5cc0, 0x19}, {0x38d70d8, 0xc001c0ee40})
                k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus/controller.go:126 +0xa5
        k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus.(*Controller).processNextWorkItem(0xc001c58380, {0x38d70d8, 0xc001c0ee40})
                k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus/controller.go:127 +0xbe
        k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus.(*Controller).runWorker(...)
                k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus/controller.go:106
        k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
                k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x1f
        k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
        k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0023dff70, {0x389ae40, 0xc003f0fce0}, 0x1, 0xc000118e00)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
        k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002914770, 0x3b9aca00, 0x0, 0x1, 0xc000118e00)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
        k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x38d70d8, 0xc001c0ee40}, 0xc00208a090, 0x3b9aca00, 0x0, 0x1)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x87
        k8s.io/apimachinery/pkg/util/wait.UntilWithContext(...)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:170
        created by k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus.(*Controller).Run in goroutine 857
                k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus/controller.go:63 +0xf8
 >
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1d4060d]
```



### What did you expect to happen?

Not to crash. By looks of it, `common.SchemaDeclType()` can (and does) return nil, and the caller at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go#L247 doesn't check the value before using it.

### How can we reproduce it (as minimally and precisely as possible)?

Probably x-kubernetes-preserve-unknown-fields extension is causing this.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
v1.32.9
```

</details>


### Cloud provider

<details>
N/A
</details>


### OS version

<details>

</details>


### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

--- a/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go
+++ b/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go
@@ -244,7 +244,11 @@ func (c *TypeChecker) declType(gvk schema.GroupVersionKind) (*apiservercel.DeclT
 	if err != nil {
 		return nil, err
 	}
-	return common.SchemaDeclType(&openapi.Schema{Schema: s}, true).MaybeAssignTypeName(generateUniqueTypeName(gvk.Kind)), nil
+	declType := common.SchemaDeclType(&openapi.Schema{Schema: s}, true)
+	if declType == nil {
+		return nil, nil
+	}
+	return declType.MaybeAssignTypeName(generateUniqueTypeName(gvk.Kind)), nil
 }
 
 func (c *TypeChecker) paramsGVK(policy *v1.ValidatingAdmissionPolicy) schema.GroupVersionKind {
@@ -404,7 +408,9 @@ func buildEnvSet(hasParams bool, hasAuthorizer bool, types typeOverwrite) (*envi
 	varOpts = append(varOpts, createVariableOpts(requestType, plugincel.RequestVarName)...)
 
 	// object and oldObject, same type, type(s) resolved from constraints
-	declTypes = append(declTypes, types.object)
+	if types.object != nil {
+		declTypes = append(declTypes, types.object)
+	}
 	varOpts = append(varOpts, createVariableOpts(types.object, plugincel.ObjectVarName, plugincel.OldObjectVarName)...)
 
 	// params, defined by ParamKind

--- a/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go
+++ b/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go
@@ -177,7 +177,7 @@ func (c *TypeChecker) CreateContext(policy *v1.ValidatingAdmissionPolicy) *TypeC
 
 func (c *TypeChecker) compiler(ctx *TypeCheckingContext, typeOverwrite typeOverwrite) (*plugincel.CompositedCompiler, error) {
 	envSet, err := buildEnvSet(
-		/* hasParams */ ctx.paramDeclType != nil,
+		/* hasParams */ !ctx.paramGVK.Empty(),
 		/* hasAuthorizer */ true,
 		typeOverwrite)
 	if err != nil {
@@ -205,7 +205,7 @@ func (c *TypeChecker) CheckExpression(ctx *TypeCheckingContext, expression strin
 			continue
 		}
 		options := plugincel.OptionalVariableDeclarations{
-			HasParams:     ctx.paramDeclType != nil,
+			HasParams:     !ctx.paramGVK.Empty(),
 			HasAuthorizer: true,
 		}
 		compiler.CompileAndStoreVariables(convertv1beta1Variables(ctx.variables), options, environment.StoredExpressions)
@@ -414,8 +414,10 @@ func buildEnvSet(hasParams bool, hasAuthorizer bool, types typeOverwrite) (*envi
 	varOpts = append(varOpts, createVariableOpts(types.object, plugincel.ObjectVarName, plugincel.OldObjectVarName)...)
 
 	// params, defined by ParamKind
-	if hasParams && types.params != nil {
-		declTypes = append(declTypes, types.params)
+	if hasParams {
+		if types.params != nil {
+			declTypes = append(declTypes, types.params)
+		}
 		varOpts = append(varOpts, createVariableOpts(types.params, plugincel.ParamsVarName)...)
 	}
 

--- a/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking_test.go
@@ -214,6 +214,24 @@ func TestTypeCheck(t *testing.T) {
 			},
 		}},
 	}}
+	noTypeSchemaPolicy := &v1.ValidatingAdmissionPolicy{Spec: v1.ValidatingAdmissionPolicySpec{
+		Validations: []v1.Validation{
+			{
+				Expression: "true",
+			},
+		},
+		MatchConstraints: &v1.MatchResources{ResourceRules: []v1.NamedRuleWithOperations{
+			{
+				RuleWithOperations: v1.RuleWithOperations{
+					Rule: v1.Rule{
+						APIGroups:   []string{"apps"},
+						APIVersions: []string{"v1"},
+						Resources:   []string{"deployments"},
+					},
+				},
+			},
+		}},
+	}}
 
 	deploymentPolicyWithBadMessageExpression := deploymentPolicy.DeepCopy()
 	deploymentPolicyWithBadMessageExpression.Spec.Validations[0].MessageExpression = "object.foo + 114514" // confusion
@@ -489,6 +507,12 @@ func TestTypeCheck(t *testing.T) {
 				},
 			},
 			assertions: []assertionFunc{toBeEmpty},
+		},
+		{
+			name:           "schema without type",
+			policy:         noTypeSchemaPolicy,
+			schemaToReturn: &spec.Schema{},
+			assertions:     []assertionFunc{toBeEmpty},
 		},
 		{
 			name: "variables valid",

--- a/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking_test.go
@@ -414,6 +414,23 @@ func TestTypeCheck(t *testing.T) {
 			},
 		},
 		{
+			name: "params with untyped schema",
+			policy: &v1.ValidatingAdmissionPolicy{Spec: v1.ValidatingAdmissionPolicySpec{
+				ParamKind: &v1.ParamKind{
+					APIVersion: "v1",
+					Kind:       "Config",
+				},
+				Validations: []v1.Validation{
+					{
+						Expression: "params != null",
+					},
+				},
+				MatchConstraints: deploymentPolicy.Spec.MatchConstraints,
+			}},
+			schemaToReturn: &spec.Schema{},
+			assertions:     []assertionFunc{toBeEmpty},
+		},
+		{
 			name:   "multiple expressions",
 			policy: multiExpressionPolicy,
 			schemaToReturn: &spec.Schema{

